driver/razerkbd_driver.c

```diff
@@ -2317,15 +2356,68 @@ static int razer_raw_event(struct hid_device *hdev, struct hid_report *report, u
             case 0x54: // Prev (song)
                 cur_value = USB_HID_KEY_MEDIA_PREVIOUSSONG;
                 break;
+            default:
+                write_bitfield = 0;
             }
 
-            data[index+1] = cur_value;
+            // data of size 22 starting with 0x01 is a bit field so we need to handle that separately
+            if (size == 22) {
+                if (write_bitfield) {
+                    if (cur_value < (sizeof(bitfield) * BITS_PER_BYTE)) {
+                        // value fits the bit field, so we can use that
+                        set_bit(cur_value, (unsigned long*) bitfield);
+                    } else {
+                        // value does not fit the bit field, so we need extra handling
+                        int report_extra = 1;
+
+                        switch (cur_value) {
+                        case USB_HID_KEY_MEDIA_VOLUMEUP:
+                            cur_value = USB_HID_USAGE_MEDIA_VOLUMEUP;
+                            break;
+                        case USB_HID_KEY_MEDIA_VOLUMEDOWN:
+                            cur_value = USB_HID_USAGE_MEDIA_VOLUMEDOWN;
+                            break;
+                        case USB_HID_KEY_MEDIA_MUTE:
+                            cur_value = USB_HID_USAGE_MEDIA_MUTE;
+                            break;
+                        case USB_HID_KEY_MEDIA_NEXTSONG:
+                            cur_value = USB_HID_USAGE_MEDIA_NEXTSONG;
+                            break;
+                        case USB_HID_KEY_MEDIA_PLAYPAUSE:
+                            cur_value = USB_HID_USAGE_MEDIA_PLAYPAUSE;
+                            break;
+                        case USB_HID_KEY_MEDIA_PREVIOUSSONG:
+                            cur_value = USB_HID_USAGE_MEDIA_PREVIOUSSONG;
+                            break;
+                        default:
+                            report_extra = 0;
+                        }
+
+                        if (report_extra) {
+                            u8 xdata[22] = { 0x02  };
+
+                            // report key down
+                            xdata[1] = cur_value;
+                            hid_report_raw_event(hdev, HID_INPUT_REPORT, xdata, sizeof(xdata), 0);
+
+                            // report key up
+                            xdata[1] = 0x00;
+                            hid_report_raw_event(hdev, HID_INPUT_REPORT, xdata, sizeof(xdata), 0);
+                        }
+                    }
+                }
+            } else { // size 16
+                data[index+1] = cur_value;
+            }
```

Comment:
I'm definitely not happy with modifying this function as I have neither a reference for it how it should work and we had too many regressions with this in the past

Reply:
Yes, that's understandable. I wasn't able to find any good documentation for this either, and I only figured the behaviour out by trial end error. It is evident however, that this change would fix multiple keyboards, as Ornata Chroma V2 had the same behaviour as BlackWidow V3 before the change (see issue #1552).

Maybe the correct approach would be to add another switch statement that differentiates by ProductID here. What do you think?

Reply:
That's the sanest solution I can think of for now. It probably makes sense to split the code for this device out into its own function and call it from here, so the function doesn't get a million lines long.